src/middleware/session.ml

```diff
@@ -355,3 +372,29 @@ let session_label request =
 
 let session_expires_at request =
   !(snd (getter request)).expires_at
+
+
+type level = Debug | Info | Success | Warning | Error
+
+let key_of_level l =
+  let suffix = match l with
+    | Debug -> "DEBUG"
+    | Info -> "INFO"
+    | Success -> "SUCCESS"
+    | Warning -> "WARNING"
+    | Error -> "ERROR" in
+  "FLASH_MESSAGE_" ^ suffix
+
+let clear_flash level request =
+  let k = key_of_level level in
+  (fst (getter request)).remove k
```

Comment:
`remove` here does not set the cookie to a time in the past, so the flash message isn't invalidated, or am I missing something?

Also to invalidate, let's not use `Session.invalidate` which calls `Unix.timeofday`, we can use `epoch` directly and save a system call

Reply:
Inside of `remove` we use `List.remove_assoc` to delete just the message associated with a particular level. Maybe the example application `b-flash-message` provides a good illustration? In that app, `/results` is supposed to show the flash message once and the user's text (stored in the session) every time.

>  let's not use Session.invalidate 

Agreed, that would wipe out everything in the session.